include/tl/adjacent.hpp

```diff
@@ -26,69 +29,76 @@ namespace tl {
 
             cursor() = default;
             constexpr cursor(cursor<!Const> i)
-                requires Const&& std::convertible_to<std::ranges::iterator_t<V>, std::ranges::iterator_t<Base>> : current_(std::move(i.current_)) {}
-
-            constexpr cursor(std::ranges::iterator_t<Base> first, std::ranges::sentinel_t<Base> last) {
-                current_[0] = first;
-                for (auto i : std::views::iota(std::size_t(1), N)) {
-                    current_[i] = std::ranges::next(current_[i - 1], 1, last);
+                requires Const && std::convertible_to<std::ranges::iterator_t<V>, std::ranges::iterator_t<Base>> :
+                first_(std::move(i.first_)),
+                last_(std::move(i.last_))
+            {}
+
+            constexpr cursor(std::ranges::iterator_t<Base> first, std::ranges::sentinel_t<Base> const last) :
+                first_(first),
+                last_(std::move(first))
+            {
+                std::ranges::advance(last_, Diff{N} - 1, last);
+                if (last_ == last) {
+                    first_ = last_;
                 }
             }
-            constexpr cursor(as_sentinel_t, std::ranges::iterator_t<Base> first, std::ranges::iterator_t<Base> last) {
-                if constexpr (!std::ranges::bidirectional_range<Base>) {
-                    for (auto& it : current_) {
-                        it = last;
-                    }
-                }
-                else {
-                    current_[N - 1] = last;
-                    for (auto i : std::views::iota(std::size_t(1), N)) {
-                        current_[N - 1 - i] = std::ranges::prev(current_[N - i], 1, first);
+            constexpr cursor(as_sentinel_t, std::ranges::iterator_t<Base> first, std::ranges::iterator_t<Base> last) :
+                first_(last),
+                last_(std::move(last))
+            {
+                if constexpr (std::ranges::bidirectional_range<Base>) {
+                    if (auto remainder = std::ranges::advance(first_, -Diff{N} + 1, first)) {
+                        first_ = last_;
                     }
                 }
             }
 
             constexpr auto read() const {
-                return tuple_transform([](auto& i) -> decltype(auto) { return *i; }, current_);
+                return [i = first_]<std::size_t... Indices>(std::index_sequence<Indices...>) mutable {
+                    using Ref = tl::meta::repeat_into<
+                        std::ranges::range_reference_t<Base>, N, detail::tuple_or_pair_impl>::type;
+                    return Ref{((void) Indices, *i++)...};
+                }(std::make_index_sequence<N>{});
```

Comment:
We could theoretically use `make_index_sequence<N - 1>` here, capture `this` instead of `first_`, and include `last_` as the final argument to `Ref`. I went with slightly more readable, but I think this suggestion may have a minor perf advantage when the cost of incrementing `Base`'s iterators is a large-ish O(1).